### DISCUSSION TITLE
Observer/observable destruction management

### DISCUSSION THREAD

**Discussion by dlyr** [STORM-IRIT/Radium-Engine#788] [Design discussion]
See beginning of the discussion here
_Originally posted by @dlyr in https://github.com/STORM-IRIT/Radium-Engine/issues/786#issuecomment-863809754_

**Comment by MathiasPaulin** [comment]
Following the bug fix in #835, here is another way to fix #834
Instead of removing  notifications on attribute deletion, we can do the following : 
- when setting up the line mesh in the function `template <typename CoreGeometry> void setupLineMesh(...)` in ForwardRenderer.cpp (lines 242 to 276), add the `RenderObject` owning the core geometry whose attrib is observed  as parameter of this function (`observedRo`)
- replace the lines 264 to 266 by : 
```c++
int verticeObserverId = core.vertexAttribs().getAttrib( handle ).attach( VerticesUpdater( disp, core ) );
ObservableVoid *observedVertices = core.vertexAttribs().getAttribPtr( handle );
// Prepare to disconnect observer at RenderObject deletion
auto signalManager = RadiumEngine::getInstance()->getSignalManager();
signalManager->getRenderObjectDestroyedNotifier().attach(
              [verticeObserverId, observedVertices, observedRo](const Scene::ItemEntry&e) {
                  if (e.m_roIndex == observedRo)
                      observedVertices->detach(verticeObserverId)  ;
               }
);
```
The lambda connected to the RenderObject destruction signal will disconnect the observers set on attribs by the function.

As SignalManager sends this signal before RO deletion (and geometry/attrib deletion), the observer will be detached before the attrib is deleted, without being notified. 
This solution follows (roughly) the previously proposed *observer detach themselves on dtor* but is only valid for RenderObject related observable. The same could be done also for Entities and Component but not for non SignalManager related observable.

I prefer the simpler solution proposed in #835 : observable should detach all observers in their destructor.
